{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import csv\n",
      "\n",
      "def get_gene_map(genome_info_fname):\n",
      "    \"\"\"Reads in the genome annotation data.\n",
      "    \n",
      "    Args:\n",
      "        genome_info_fname: filename for the tab-delimited genome\n",
      "            info file.\n",
      "    \n",
      "    Returns:\n",
      "        A dictionary mapping the locusId to a dictionary of \n",
      "        gene annotations.\n",
      "    \"\"\"\n",
      "    genes_data = {}\n",
      "    with open(genome_info_fname) as inf:\n",
      "        # Dict reader reads a tabular file and returns\n",
      "        # each row as a dictionary mapping rows headers to values\n",
      "        # for that row.\n",
      "        reader = csv.DictReader(inf, dialect='excel-tab')\n",
      "        for row in reader:\n",
      "            # Store the gene info by locus ID genes_data.\n",
      "            gene_name = row['locusId']\n",
      "            genes_data[gene_name] = row\n",
      "    # genes_data is now initialized.\n",
      "    return genes_data\n",
      "\n",
      "\n",
      "def get_operon_list(operon):\n",
      "    \"\"\"Reads in the operon annotation data.\n",
      "    \n",
      "    Args:\n",
      "        operon_pred_fname: filename for operon predictions.\n",
      "    \n",
      "    Returns:\n",
      "        A list of sets of locus IDs of all operons containing \n",
      "        more than 1 gene.\n",
      "    \"\"\"\n",
      "    # General strategy: Keep a list of sets of genes that are together\n",
      "    # in operons. Also keep a dictionary mapping locus IDs to indices into\n",
      "    # that list. If two genes are in the same operon, they will have the \n",
      "    # same index in the dictionary. This allows us to find the operon from\n",
      "    # the locus ID of any gene that is already in it.\n",
      "    operon_list = [] # list of operons\n",
      "    operon_map = {}  # maps gene_names to indexes in operon_list\n",
      "    with open('gnc511145.named') as inf:\n",
      "        # Another DictReader, like above.\n",
      "        reader = csv.DictReader(inf, dialect='excel-tab')\n",
      "        \n",
      "        operon_genes = set()\n",
      "        for row in reader:\n",
      "            gene_a = row['Gene1'].strip()\n",
      "            gene_b = row['Gene2'].strip()\n",
      "            # Names are valid if both evaluate to True, i.e. neither is empty.\n",
      "            valid_names = gene_a and gene_b\n",
      "            # bOp is a prediction of whether both are in an operon\n",
      "            in_operon = bool(row['bOp'])  \n",
      "            \n",
      "            # Consider something in an operon only if both \n",
      "            # genes have valid (non-empty) names. \n",
      "            # Since we use the locus ID for the name, I believe all of them \n",
      "            # are valid. But this is called \"defensive programming.\"\n",
      "            in_operon = in_operon and valid_names\n",
      "            \n",
      "            # Index of the operon set \n",
      "            idx = operon_map.get(gene_a) or operon_map.get(gene_b)\n",
      "            \n",
      "            # Current pair not in an operon\n",
      "            if in_operon:\n",
      "                if idx is None:\n",
      "                    idx = len(operon_list)\n",
      "                    operon_list.append(set([gene_a, gene_b]))\n",
      "                else:\n",
      "                    operon_list[idx].update(set([gene_a, gene_b]))\n",
      "                operon_map[gene_a] = idx\n",
      "                operon_map[gene_b] = idx\n",
      "    return operon_list \n",
      "\n",
      "\n",
      "# Template for an operon row.\n",
      "OPERON_ROW_KEYS = ['seqname', 'source', 'feature', 'start',\n",
      "                   'end', 'score', 'strand', 'frame', 'attributes']\n",
      "OPERON_ROW_TEMPLATE = {'seqname': 'Chromosome',\n",
      "                       'source': 'MicrobesOnline',\n",
      "                       'feature': 'operon',\n",
      "                       'score': '.',\n",
      "                       'frame': '.'}\n",
      "def make_operon_row(strand, start, stop, attributes):\n",
      "    \"\"\"Returns a dictionary representing the operon.\"\"\"\n",
      "    # Copy the default operon row and modify.\n",
      "    op_data = OPERON_ROW_TEMPLATE.copy()    \n",
      "    op_data['strand'] = strand\n",
      "    op_data['start'] = start\n",
      "    op_data['end'] = stop\n",
      "    op_data['attributes'] = attributes\n",
      "    return op_data"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_map = get_gene_map('genomeInfo.tsv')\n",
      "operon_list = get_operon_list('gnc511145.scores.txt')\n",
      "all_genes_in_operons = set()\n",
      "for op in operon_list:\n",
      "    all_genes_in_operons.update(op)\n",
      "\n",
      "# Make this GFF file now. \n",
      "# This first loop only covers genes in multi-gene operons.\n",
      "operon_rows = []\n",
      "for op in operon_list:\n",
      "    names = [gene_map[n]['name'] for n in op]\n",
      "    op_name = '-'.join(sorted(names))\n",
      "    strands = set([gene_map[n]['strand'] for n in op])\n",
      "    if len(strands) != 1:\n",
      "        print 'Operon genes not on same strand, skipping'\n",
      "        continue\n",
      "        \n",
      "    strand = strands.pop()\n",
      "    \n",
      "    starts = [int(gene_map[n]['start']) for n in op]\n",
      "    stops = [int(gene_map[n]['stop']) for n in op]\n",
      "    if strand == '+':\n",
      "        # If we are on the + strand, the start of the operon\n",
      "        # is the first start position of any gene and the end\n",
      "        # is the last stop of any gene.\n",
      "        start = min(starts)\n",
      "        stop = max(stops)\n",
      "    else:\n",
      "        # If we are on the - strand, numbers are reverse.\n",
      "        start = max(starts)\n",
      "        stop = min(stops)\n",
      "    \n",
      "    attr = r'operon \"%s\"' % op_name \n",
      "    op_data = make_operon_row(strand, start, stop, attr)\n",
      "    operon_rows.append(op_data)\n",
      "\n",
      "    \n",
      "# Notice that we omitted genes that are not in operons.\n",
      "# Let's make operons for all the genes we missed.\n",
      "all_genes = set(gene_map.keys())\n",
      "genes_not_in_operons = all_genes.difference(all_genes_in_operons)\n",
      "for gene in genes_not_in_operons:\n",
      "    d = gene_map[gene]\n",
      "    start = int(d['start'])\n",
      "    stop = int(d['stop'])\n",
      "    strand = d['strand']\n",
      "    attributes = r'operon \"%s\"' % d['name']\n",
      "    op_data = make_operon_row(strand, start, stop, attr)\n",
      "    operon_rows.append(op_data)\n",
      "\n",
      "# Order in which to write the fields to the file. \n",
      "with open('GFFoutput.gff', 'w') as outf:\n",
      "    # OPERON_ROW_KEYS is defined above and determines the order\n",
      "    # in which the DictWriter writes the fields in the dictionary.\n",
      "    writer = csv.DictWriter(outf, OPERON_ROW_KEYS, dialect='excel-tab',\n",
      "                            quoting=csv.QUOTE_NONE,\n",
      "                            escapechar='\\\\')\n",
      "    writer.writerows(operon_rows)  # Writes all the rows."
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print 'Printing GFF output:'\n",
      "print\n",
      "with open('GFFoutput.gff') as gff:\n",
      "    for i, line in enumerate(gff):\n",
      "        print line.strip()\n",
      "        if i > 10: break"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Printing GFF output:\n",
        "\n",
        "Chromosome\tMicrobesOnline\toperon\t190\t2799\t.\t+\t.\toperon \\\"thrA-thrL\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t337\t5530\t.\t+\t.\toperon \\\"thrA-thrB-thrC-yaaX\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t7959\t5683\t.\t-\t.\toperon \\\"yaaA-yaaJ\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t8238\t9893\t.\t+\t.\toperon \\\"mog-talB\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t11786\t9928\t.\t-\t.\toperon \\\"yaaH-yaaI-yaaW\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t12163\t16557\t.\t+\t.\toperon \\\"dnaJ-dnaK-insL-1\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t16960\t16751\t.\t-\t.\toperon \\\"hokC-mokC\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t16952\t19620\t.\t+\t.\toperon \\\"nhaA-nhaR-sokC\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t21078\t19811\t.\t-\t.\toperon \\\"insA-1-insB-1-rpsT\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t21181\t34695\t.\t+\t.\toperon \\\"caiF-carA-carB-dapB-fkpB-ileS-ispH-lspA-ribF-rihC-yaaY\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t41931\t34781\t.\t-\t.\toperon \\\"caiA-caiB-caiC-caiD-caiE-caiT\\\"\n",
        "Chromosome\tMicrobesOnline\toperon\t42403\t50302\t.\t+\t.\toperon \\\"fixA-fixB-fixC-fixX-folA-kefC-kefF-yaaU\\\"\n"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}